src/main.cpp

```diff
@@ -1284,17 +1403,33 @@ bool CheckTransactionWithoutProofVerification(const CTransaction& tx, CValidatio
                             REJECT_INVALID, "bad-tx-expiry-height-too-high");
         }
     }
+    auto orchard_bundle = tx.GetOrchardBundle();
 
-    // Transactions containing empty `vin` must have either non-empty
-    // `vJoinSplit` or non-empty `vShieldedSpend`.
-    if (tx.vin.empty() && tx.vJoinSplit.empty() && tx.vShieldedSpend.empty())
+    // Transactions must contain some potential source of funds. This
+    // rejects obviously-invalid transaction constructions early, but
+    // cannot prevent e.g. a pure Sapling transaction with only dummy
+    // spends (which is undetectable).
+    if (tx.vin.empty() &&
+        tx.vJoinSplit.empty() &&
+        tx.vShieldedSpend.empty() &&
+        !orchard_bundle.SpendsEnabled())
+    {
         return state.DoS(10, error("CheckTransaction(): vin empty"),
                          REJECT_INVALID, "bad-txns-vin-empty");
-    // Transactions containing empty `vout` must have either non-empty
-    // `vJoinSplit` or non-empty `vShieldedOutput`.
-    if (tx.vout.empty() && tx.vJoinSplit.empty() && tx.vShieldedOutput.empty())
+    }
+    // Transactions must contain some potential useful sink of funds.
+    // This rejects obviously-invalid transaction constructions early,
+    // but cannot prevent e.g. a pure Sapling transaction with only
+    // dummy outputs (which is undetectable), and does not prevent
+    // transparent transactions from sending all funds to miners.
+    if (tx.vout.empty() &&
+        tx.vJoinSplit.empty() &&
+        tx.vShieldedOutput.empty() &&
+        !orchard_bundle.OutputsEnabled())
```

Comment:
This doesn't seem to implement the specified consensus rule, because it doesn't check `nActionsOrchard`.

Reply:
I should add a comment - these calls that look like flag checks are confusing, because they also check that the Orchard bundle is present. If an Orchard bundle exists, its actions vector is nonempty by construction.